### PR TITLE
Added reset password support

### PR DESCRIPTION
**Pull request by bastien70**

### CODE REVIEW

File: config/packages/twig.yaml [hunk 1]
```diff
@@ -1,5 +1,7 @@
 twig:
     default_path: '%kernel.project_dir%/templates'
+    form_themes:
+        - "bootstrap_5_layout.html.twig"
```
Comment (jmsche) [inline config/packages/twig.yaml:4]:
```suggestion
        - 'bootstrap_5_layout.html.twig'
```
Using simple quotes to be consistent with the previous line

File: migrations/Version20210903135827.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace DoctrineMigrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20210903135827 extends AbstractMigration
```
Comment (jmsche) [inline migrations/Version20210903135827.php:13]:
```suggestion
final class Version20210903135827 extends AbstractMigration
```

File: src/Controller/ResetPasswordController.php [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        if ($form->isSubmitted() && $form->isValid()) {
+            return $this->processSendingPasswordResetEmail(
+                $resetPasswordRequest->email,
+                $mailer
+            );
+        }
+
+        return $this->render('reset_password/request.html.twig', [
+            'requestForm' => $form->createView(),
+        ]);
```
Comment (jmsche) [inline src/Controller/ResetPasswordController.php:56]:
```suggestion
        return $this->renderForm('reset_password/request.html.twig', [
            'requestForm' => $form,
        ]);
```

File: src/Controller/ResetPasswordController.php [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+        try {
+            /** @var User $user */
+            $user = $this->resetPasswordHelper->validateTokenAndFetchUser($token);
+        } catch (ResetPasswordExceptionInterface $e) {
+            $this->addFlash('error', sprintf(
+                '%s - %s',
+                new TranslatableMessage('reset_password.reset.flash_error'),
+                $e->getReason()
+            ));
```
Comment (jmsche) [inline src/Controller/ResetPasswordController.php:103]: Did you test that it was correctly translated?

File: src/Controller/ResetPasswordController.php [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                'success',
+                new TranslatableMessage('reset_password.reset.flash_success')
+            );
+
+            return $this->redirectToRoute('app_login');
+        }
+
+        return $this->render('reset_password/reset.html.twig', [
+            'resetForm' => $form->createView(),
+        ]);
```
Comment (jmsche) [inline src/Controller/ResetPasswordController.php:139]:
```suggestion
        return $this->renderForm('reset_password/reset.html.twig', [
            'resetForm' => $form,
        ]);
```

File: src/Controller/ResetPasswordController.php [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+        // Do not reveal whether a user account was found or not.
+        if (!$user) {
+            return $this->redirectToRoute('app_check_email');
+        }
+
+        try {
+            $resetToken = $this->resetPasswordHelper->generateResetToken($user);
+        } catch (ResetPasswordExceptionInterface $e) {
```
Comment (jmsche) [inline src/Controller/ResetPasswordController.php:156]:
```suggestion
        } catch (ResetPasswordExceptionInterface) {
```

File: src/Entity/ResetPasswordRequest.php [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    public function __construct(User $user, \DateTimeInterface $expiresAt, string $selector, string $hashedToken)
+    {
+        $this->user = $user;
+        $this->initialize($expiresAt, $selector, $hashedToken);
+    }
+
+    public function getId(): ?int
+    {
+        return $this->id;
+    }
```
Comment (jmsche) [inline src/Entity/ResetPasswordRequest.php:32]: Isn't this already in the PrimaryKeyTrait?

File: src/Form/Model/ResetPasswordRequestModel.php [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+declare(strict_types=1);
+
+namespace App\Form\Model;
+
+use Symfony\Component\Validator\Constraints as Assert;
+
+final class ResetPasswordRequestModel
+{
+    #[Assert\NotBlank()]
```
Comment (jmsche) [inline src/Form/Model/ResetPasswordRequestModel.php:11]: I think adding an Email assert as well would be great

File: templates/bundles/EasyAdminBundle/login_layout.html.twig [hunk 9]
```diff
@@ -0,0 +1,51 @@
+{# @var ea \EasyCorp\Bundle\EasyAdminBundle\Context\AdminContext #}
```
Comment (jmsche) [inline templates/bundles/EasyAdminBundle/login_layout.html.twig:1]: Why duplicating all this code?

File: templates/bundles/EasyAdminBundle/page/login.html.twig [hunk 10]
```diff
@@ -0,0 +1,38 @@
+{% extends 'bundles/EasyAdminBundle/login_layout.html.twig' %}
```
Comment (jmsche) [inline templates/bundles/EasyAdminBundle/page/login.html.twig:1]: Why duplicating all this code?

File: templates/reset_password/email.html.twig [hunk 11]
```diff
@@ -0,0 +1,27 @@
+{#<h1>Hi!</h1>#}
+
+{#<p>To reset your password, please visit the following link</p>#}
+
+{#<a href="{{ url('app_reset_password', {token: resetToken.token}) }}">{{ url('app_reset_password', {token: resetToken.token}) }}</a>#}
+
+{#<p>This link will expire in {{ resetToken.expirationMessageKey|trans(resetToken.expirationMessageData, 'ResetPasswordBundle') }}.</p>#}
+
+{#<p>Cheers!</p>#}
+
```
Comment (jmsche) [inline templates/reset_password/email.html.twig:10]: I think comments can be removed

File: tests/Controller/ResetPasswordControllerTest.php [hunk 12]
```diff
@@ -0,0 +1,20 @@
+<?php
```
Comment (jmsche) [inline tests/Controller/ResetPasswordControllerTest.php:1]: Maybe more tests can be added?

**Review by jmsche — CHANGES_REQUESTED** [review]

File: templates/bundles/EasyAdminBundle/login_layout.html.twig (hunk 9 above)
Comment (bastien70) [reply templates/bundles/EasyAdminBundle/login_layout.html.twig:1]: This is a layout that I created myself which will take the main elements of the login page. It serves as a basic template for the other pages (request reset password and reset password), I just have to overload the content block by adding the form. This basic template also contains support for flash messages.

File: templates/bundles/EasyAdminBundle/page/login.html.twig (hunk 10 above)
Comment (bastien70) [reply templates/bundles/EasyAdminBundle/page/login.html.twig:1]: See https://github.com/bastien70/dbsaver/pull/36#discussion_r702439034

File: src/Controller/ResetPasswordController.php (hunk 4 above)
Comment (bastien70) [reply src/Controller/ResetPasswordController.php:103]: I updated the way to handle this. The translation will be taken into account. Unfortunately, the error message ($ e-> getReason ()) will remain in English no matter what. Possibly we could add a bundle allowing to translate text in any language, but beware of word-for-word translation errors

File: templates/bundles/EasyAdminBundle/page/login.html.twig (hunk 10 above)
Comment (jmsche) [reply templates/bundles/EasyAdminBundle/page/login.html.twig:1]: Not needed here I think, see v3.5.8 of EA

File: tests/Controller/ResetPasswordControllerTest.php (hunk 12 above)
Comment (bastien70) [reply tests/Controller/ResetPasswordControllerTest.php:1]: I would like to test the reset password itself, but when I manually call the ResetPasswordHelperInterface :: generateResetToken method, I always get the TooManyPasswordRequestsException exception in test mode, which prevents me from going any further. I can't figure out how to get by at this level

File: templates/bundles/EasyAdminBundle/page/login.html.twig (hunk 10 above)
Comment (bastien70) [reply templates/bundles/EasyAdminBundle/page/login.html.twig:1]: It doesn't take into account custom errors if I'm not mistaken. Hence the fact that I added: 

`{% include '@EasyAdmin/flash_messages.html.twig' %}`

File: tests/Controller/ResetPasswordControllerTest.php (hunk 12 above)
Comment (jmsche) [reply tests/Controller/ResetPasswordControllerTest.php:1]: Maybe change the throttle_limit to a short value for tests (eg. 1 second)

File: templates/bundles/EasyAdminBundle/page/login.html.twig (hunk 10 above)
Comment (jmsche) [reply templates/bundles/EasyAdminBundle/page/login.html.twig:1]: The login page itself does not need to be overridden anymore as the login template from EA now contains "forgot password" link.

Comment (jmsche) [reply templates/bundles/EasyAdminBundle/page/login.html.twig:1]: Okay, I've just seen why it's required to override it even with 3.5.8.

File: src/Controller/ResetPasswordController.php (hunk 4 above)
Comment (jmsche) [reply src/Controller/ResetPasswordController.php]: I think you could try to catch specific exceptions and then have a specific translated reasons?

File: tests/Controller/ResetPasswordControllerTest.php [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    private UserRepository $userRepository;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+        $this->resetPasswordHelper = self::getContainer()->get(ResetPasswordHelperInterface::class);
+        $this->userRepository = self::getContainer()->get(UserRepository::class);
+    }
+
+    public function testRequestPasswordReset()
```
Comment (jmsche) [inline tests/Controller/ResetPasswordControllerTest.php:24]:
```suggestion
    public function testRequestPasswordReset(): void
```

File: tests/Controller/ResetPasswordControllerTest.php [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+        ]]);
+
+        self::assertResponseRedirects('/reset-password/check-email');
+    }
+
+    /**
+     * @throws ResetPasswordExceptionInterface
+     * @throws TooManyPasswordRequestsException
+     */
+    public function testResetPassword()
```
Comment (jmsche) [inline tests/Controller/ResetPasswordControllerTest.php:40]:
```suggestion
    public function testResetPassword(): void
```

File: translations/messages.en.yaml [hunk 15]
```diff
@@ -124,3 +124,24 @@ user:
...
+        expires_message: This link will expire in %expires_in%.
+        not_received: If you don't receive an email please check your spam folder or <a href="%app_forgot_password_request_route%">try again</a>.
+    reset:
+        title: Reset your password
+        submit: Reset
+        field:
+            new_password: New password
+            repeat_password: Repeat Password
+        flash_success: 'Your password has been reset! You can now log in.'
+        flash_error: 'A problem occurred while validating your reset request : %message%'
```
Comment (jmsche) [inline translations/messages.en.yaml:147]:
```suggestion
        flash_error: 'A problem occurred while validating your reset request: %message%.'
```

File: translations/messages.fr.yaml [hunk 16]
```diff
@@ -124,3 +124,24 @@ user:
...
+        expires_message: Ce lien va expirer dans %expires_in%.
+        not_received: Si vous ne recevez pas d'email, veuillez vérifier votre dossier spam ou <a href="%app_forgot_password_request_route%">réessayez</a>.
+    reset:
+        title: Réinitialiser mon mot de passe
+        submit: Réinitialiser
+        field:
+            new_password: Nouveau mot de passe
+            repeat_password: Répétez le mot de passe
+        flash_success: 'Votre mot de passe a bien été réinitialisé ! Vous pouvez désormais vous connecter.'
+        flash_error: 'Un problème est survenu lors de la validation de votre demande de réinitialisation : %message%'
```
Comment (jmsche) [inline translations/messages.fr.yaml:147]:
```suggestion
        flash_error: 'Un problème est survenu lors de la validation de votre demande de réinitialisation : %message%.'
```

**Review by jmsche — CHANGES_REQUESTED** [review]

File: src/Controller/ResetPasswordController.php [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+use Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface;
+use Symfony\Component\Routing\Annotation\Route;
+use Symfony\Component\Translation\TranslatableMessage;
+use Symfony\Contracts\Translation\TranslatorInterface;
+use SymfonyCasts\Bundle\ResetPassword\Controller\ResetPasswordControllerTrait;
+use SymfonyCasts\Bundle\ResetPassword\Exception\ResetPasswordExceptionInterface;
+use SymfonyCasts\Bundle\ResetPassword\ResetPasswordHelperInterface;
+
+#[Route('/reset-password')]
+class ResetPasswordController extends AbstractController
```
Comment (jmsche) [inline src/Controller/ResetPasswordController.php:27]:
```suggestion
final class ResetPasswordController extends AbstractController
```

File: tests/Controller/ResetPasswordControllerTest.php [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+declare(strict_types=1);
+
+namespace App\Tests\Controller;
+
+use App\Repository\UserRepository;
+use SymfonyCasts\Bundle\ResetPassword\Exception\ResetPasswordExceptionInterface;
+use SymfonyCasts\Bundle\ResetPassword\Exception\TooManyPasswordRequestsException;
+use SymfonyCasts\Bundle\ResetPassword\ResetPasswordHelperInterface;
+
+class ResetPasswordControllerTest extends AbstractControllerTest
```
Comment (jmsche) [inline tests/Controller/ResetPasswordControllerTest.php:12]:
```suggestion
final class ResetPasswordControllerTest extends AbstractControllerTest
```

File: tests/Controller/ResetPasswordControllerTest.php [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+            'email' => 'user@test.com',
+        ]]);
+
+        self::assertResponseRedirects('/reset-password/check-email');
+    }
+
+    /**
+     * @throws ResetPasswordExceptionInterface
+     * @throws TooManyPasswordRequestsException
+     */
```
Comment (jmsche) [inline tests/Controller/ResetPasswordControllerTest.php:39]: Necessary? Should not throw anything

**Review by jmsche — CHANGES_REQUESTED** [review]

**Review by jmsche — APPROVED** [review]

**Pull request merged**